### ISSUE TITLE
Laravel 9 compatibility

### ISSUE THREAD

**Issue by luchaos** [commercetools/commercetools-sdk-php-v2#74]
**Is your feature request related to a problem? Please describe.**

Cannot update to an installable set of dependencies when upgrading to Laravel 9 and any version of the SDK.

The [SDK depends on `cache/filesystem-adapter`](https://github.com/commercetools/commercetools-sdk-php-v2/blob/master/composer.json#L64) which in turn [depends on `league/flysystem` v1](https://github.com/php-cache/cache/blob/master/composer.json#L26). Laravel depends on [`league/flysystem` v3](https://github.com/laravel/framework/blob/9.x/composer.json#L27) ([Upgrade Guide](https://laravel.com/docs/9.x/upgrade#flysystem-3) for reference).

**Describe the solution you'd like**

Bringing one's own cache should not interfere with other optional dependencies that might not be used.

`league/flysystem` is a very popular, low-level package - chances of running into conflicts are high.

As a PSR 6 compliant cache instance may be [passed into the SDK](https://github.com/commercetools/commercetools-sdk-php-v2/blob/9625487b3b1efe25ed86116060c6b8a4c1e8cffe/lib/commercetools-base/src/Client/OAuthHandlerFactory.php#L29) (which we do - we inject Laravel's cache) I'd suggest making the installation of `cache/filesystem-adapter` optional.
The reasoning being that the use of an internal fallback is optional by default thus the dependency itself should be optional as well.

- Make `cache/filesystem-adapter` "just" a suggested package in a new major version
- Check for its existence before [instantiation](https://github.com/commercetools/commercetools-sdk-php-v2/blob/9625487b3b1efe25ed86116060c6b8a4c1e8cffe/lib/commercetools-base/src/Client/OAuthHandlerFactory.php#L32-L34)
- Throw an exception when no cache was given nor an expected adapter installed
- Additionally point out that requirement in the [installation docs/readme](https://docs.commercetools.com/sdk/php-sdk#authentication).

💡 Mostly future proof but will require additional dependencies after installation (or injecting another cache instance as is supported now already). Very much the desirable outcome in my opinion (unless there's a much simpler yet robust solution to this, obviously).

**Describe alternatives you've considered**

Wait for [`cache/filesystem-adapter` PR to be merged](https://github.com/php-cache/cache/pull/265) (which will add `league/flysystem` v3 support), then update the SDK's dependency on `cache/filesystem-adapter` accordingly.

💡 Not as future proof, might break again, but will work out of the box without any additional dependencies after installation (or providing another cache instance, again). The easy way out, no estimated time of arrival.

**Additional Information**

[php-cache/cache](https://github.com/php-cache/cache):
> Back in 2016, this was the first library supporting PSR-6. We created Symfony bundles and made many great libraries in the PHP-cache organization. This was a few years ago and the library is not activly maintained. Starting a new project one should consider using the more performant and activly supported [Symfony Cache](https://symfony.com/doc/current/components/cache.html)[](https://github.com/php-cache/cache#notice).

**Steps to reproduce**

```shell
composer create-project laravel/laravel laravel-commercetools
cd laravel-commercetools
composer require commercetools/commercetools-sdk  -W
```

```
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - cache/filesystem-adapter[1.1.0, ..., 1.2.0] require league/flysystem ^1.0 -> found league/flysystem[1.0.0-alpha1, ..., 1.x-dev] but these were not loaded, likely because it conflicts with another require.
    - commercetools/commercetools-sdk 4.4.0 requires cache/filesystem-adapter ^1.1.0 -> satisfiable by cache/filesystem-adapter[1.1.0, 1.1.x-dev, 1.2.0].
    - Root composer.json requires commercetools/commercetools-sdk ^4.4 -> satisfiable by commercetools/commercetools-sdk[4.4.0].
```

**Comment by luchaos** [comment]
Step to reproduce:
```shell
composer create-project laravel/laravel laravel-commercetools
cd laravel-commercetools
composer require commercetools/commercetools-sdk  -W
```

```
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - cache/filesystem-adapter[1.1.0, ..., 1.2.0] require league/flysystem ^1.0 -> found league/flysystem[1.0.0-alpha1, ..., 1.x-dev] but these were not loaded, likely because it conflicts with another require.
    - commercetools/commercetools-sdk 4.4.0 requires cache/filesystem-adapter ^1.1.0 -> satisfiable by cache/filesystem-adapter[1.1.0, 1.1.x-dev, 1.2.0].
    - Root composer.json requires commercetools/commercetools-sdk ^4.4 -> satisfiable by commercetools/commercetools-sdk[4.4.0].
```

**Comment by luchaos** [comment]
Having forked the repository just now I realized that you, @jenschude , already thought of this here: https://github.com/commercetools/commercetools-sdk-php-v2/commit/50943e025a15f26266b265883ad1bdda4bf899f8

Thanks! Switching to that branch works just fine for now 👍  All it took was a `composer require commercetools/commercetools-sdk:dev-filesystem-cache-change`. 
While I can see how that's probably not the most stable nor pretty solution it sure helped me continuing with the Laravel 9 upgrade for now. I hope to see that change in one of the next releases 🙇 

I'll leave that issue to your consideration then.

**Comment by lchhieu** [comment]
The same isssue when intergtate in laravel 9. Any update?

**Comment by jenschude** [comment]
There was a release 13 days ago where we changed to the symfony cache component. This should fix the issue

**Comment by lchhieu** [comment]
> 

Hi @jenschude But this commit have reverted your changes. https://github.com/commercetools/commercetools-sdk-php-v2/commit/ee306baf442855da0e73dadfc7bcf35a324f35e8#diff-475044f06efea75a0f624e0e9d694c07619e9820df701bd572dd4c704760de4d

https://github.com/commercetools/commercetools-sdk-php-v2/blob/master/lib/commercetools-base/src/Client/OAuthHandlerFactory.php

**Comment by jenschude** [comment]
I will take care of it and do a patch release latest tomorrow

**Comment by jenschude** [comment]
Oh man. Forgot to add the changes to the code generator. So the next run reverted the changes. I just released the 5.0.1

**Comment by andreintm** [comment]
hey, any updates on that? Still same issue with `cache/filesystem-adapter` that is on V1.

**Comment by jenschude** [comment]
Just released 7.1.1

In case it's needed in a previous version. Please ping here so I can backport it.

And thanks for the hint.
